### ISSUE TITLE
[FR] Add exclude parameter to download_artifacts() and mlflow artifacts download CLI to skip unwanted artifact subdirectories.

### ISSUE THREAD

**Issue by kmohee** [mlflow/mlflow#22097]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Willingness to contribute

Yes. I can contribute this feature independently.

## Proposal Summary
Add an `exclude` parameter to `mlflow.artifacts.download_artifacts()`  and an `--exclude` flag to the `mlflow artifacts download` CLI command,  allowing users to skip specific subdirectories or file patterns during artifact download.

## Motivation

When downloading a model logged with frameworks like PyTorch Lightning, training checkpoints are often stored as a subdirectory alongside the model weights (e.g. `model/checkpoints/`). These checkpoints can be significantly larger than the model weights themselves  (often by an order of magnitude), but are not required for inference or serving.

Currently, `download_artifacts()` downloads the entire artifact directory with no way to selectively exclude subdirectories or file patterns. The only workaround is to know the exact path to the weights file and construct a specific `artifact_path`, which is brittle and undocumented.

### Example of the problem
```python
# This downloads model weights AND all training checkpoints
mlflow.artifacts.download_artifacts(
    artifact_uri="models:/my_model/Production"
)

# Workaround: brittle and requires knowing internal structure
mlflow.artifacts.download_artifacts(
    artifact_uri="models:/my_model/Production",
    artifact_path="data/model.pth"  # not always obvious or stable
)
```

This issue is related to but distinct from #5241, which proposes downloading *more* artifacts. This proposal is about downloading *less*.

### Python API
```python
mlflow.artifacts.download_artifacts(
    artifact_uri="models:/my_model/Production",
    exclude=["checkpoints/", "*.ckpt"]  # glob patterns or paths
)
```

### CLI
```bash
mlflow artifacts download \
  --artifact-uri models:/my_model/Production \
  --exclude "checkpoints/" \
  --exclude "*.ckpt"
```

The `exclude` parameter would accept a list of glob-style patterns. Any artifact whose relative path matches a pattern would be skipped during download. 

## Implementation Notes

The implementation would touch:
- `mlflow/artifacts/__init__.py` — add `exclude` parameter to `download_artifacts()`
- `mlflow/store/artifact/artifact_repo.py` — filter artifact paths before download in `ArtifactRepository.download_artifacts()`
- `mlflow/cli.py` — add `--exclude` option to the `mlflow artifacts download` command
- Appropriate unit tests for pattern matching and filtering behaviour

## Related Issues
- #5241 — complementary proposal for downloading more artifacts via model URI

### Motivation

> #### What is the use case for this feature?
A user downloading a model for deployment should not be forced to pull gigabytes of optimizer states and intermediate checkpoints they will never use.

> #### Why is this use case valuable to support for MLflow users in general?
Any team running regular training jobs with checkpoint callbacks will accumulate large checkpoint directories in their artifact store. Every downstream consumer: serving pipelines, CI/CD workflows, evaluation scripts, that calls download_artifacts() pays the full bandwidth and storage cost even when only the final weights are needed. This is especially painful in cloud environments where egress costs money and download time directly impacts deployment latency.

> #### Why is this use case valuable to support for your project(s) or organization?
In production ML pipelines, model promotion and deployment are time-sensitive. Downloading unnecessary artifacts adds latency to every deployment cycle. Teams working with large models (multi-GB weights) where checkpoints may be 5-10x the size of the final weights feel this cost acutely on every rollout.

> #### Why is it currently difficult to achieve this use case?
The only current workaround is to know the exact internal path to the weights file and pass it as artifact_path — for example artifact_path="data/model.pth". This is brittle because the internal structure varies by framework and MLflow flavour, is not documented as a stable interface, and breaks silently if the structure changes between framework versions. There is no way to express "give me everything except this subdirectory" without writing custom post-download cleanup code.

### What machine learning domain(s) is this feature request about?

- [ ] `domain/genai`: LLMs, Agents, and other GenAI-related use cases
- [ ] `domain/classical-ml`: Traditional machine learning, such as linear regression.
- [ ] `domain/deep-learning`: Deep learning and neural networks.
- [x] `domain/platform`: MLflow platform foundation, not specific to a particular machine learning domain.

### What area(s) of MLflow is this feature request about?

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [x] `area/docs`: MLflow documentation pages

**Comment by TomeHirata** [comment]
Thank you for the report! Can I know how you log your model and checkpoints? Ideally, checkpoints should be logged to the run artifacts and not included in the model artifacts so that `download_artifacts` doesn't load them when `models:/` uri is used.

**Comment by kmohee** [comment]
Thanks for the clarification! I've reproduced both cases. You're correct that models:/ correctly excludes checkpoints. However, runs:/ downloads the full artifact directory including all checkpoints.

The --exclude flag would still be valuable for users who:
1. Haven't registered their model in the registry
2. Use runs:/ for quick ad-hoc downloads
3. Work with framework integrations (e.g. Lightning) where checkpoint 
   placement inside model/ isn't directly in their control

That said, I'm happy to take your advice on whether this is worth addressing, if the recommended path is simply to use models:/ and register models, I understand. Let me know how you'd like to proceed.

**Comment by TomeHirata** [comment]
In MLflow 3, `models:/` should be available even before you register your models in the registry so using the uri for loading model weights and config only is our recommendation. I'll close this issue for now, but please feel free to reopen it if you find model:/ is not sufficient for some workloads.
https://mlflow.org/docs/latest/ml/mlflow-3/#model-artifacts-storage-location
